src/Compilers/Core/Portable/Compilation/DeterministicKeyBuilder.cs

```diff
@@ -0,0 +1,529 @@
+                    writer.Write("compilerVersion", compilerVersion);
+
+                    var runtimeVersion = typeof(object).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()?.InformationalVersion;
+                    writer.Write("runtimeVersion", runtimeVersion);
+
+                    writer.Write("framework", RuntimeInformation.FrameworkDescription);
+                    writer.Write("os", RuntimeInformation.OSDescription);
+                }
+
+                writer.WriteObjectEnd();
+            }
+        }
+
+        private void WriteSyntaxTree(
+            JsonWriter writer,
+            SyntaxTreeKey syntaxTree,
+            ImmutableArray<KeyValuePair<string, string>> pathMap,
+            DeterministicKeyOptions options,
+            CancellationToken cancellationToken)
+        {
+            writer.WriteObjectStart();
+            WriteFilePath(writer, "fileName", syntaxTree.FilePath, pathMap, options);
+            writer.WriteKey("text");
+            WriteSourceText(writer, syntaxTree.GetText(cancellationToken));
+            writer.WriteKey("parseOptions");
+            WriteParseOptions(writer, syntaxTree.Options);
+            writer.WriteObjectEnd();
+        }
+
+        private void WriteSourceText(JsonWriter writer, SourceText? sourceText)
+        {
+            if (sourceText is null)
+            {
+                return;
+            }
+
+            writer.WriteObjectStart();
+            WriteByteArrayValue(writer, "checksum", sourceText.GetChecksum().AsSpan());
+            writer.Write("checksumAlgorithm", sourceText.ChecksumAlgorithm);
+            writer.Write("encoding", sourceText.Encoding?.EncodingName);
+            writer.WriteObjectEnd();
+        }
+
+        internal void WriteMetadataReference(
+            JsonWriter writer,
+            MetadataReference reference,
+            DeterministicKeyOptions deterministicKeyOptions,
+            CancellationToken cancellationToken)
+        {
+            writer.WriteObjectStart();
+            if (reference is PortableExecutableReference peReference)
+            {
+                switch (peReference.GetMetadata())
+                {
+                    case AssemblyMetadata assemblyMetadata:
+                        {
+                            var modules = assemblyMetadata.GetModules();
+                            writeModuleMetadata(modules[0]);
+                            writer.WriteKey("secondaryModules");
+                            writer.WriteArrayStart();
+                            for (var i = 1; i < modules.Length; i++)
+                            {
+                                writer.WriteObjectStart();
+                                writeModuleMetadata(modules[i]);
+                                writer.WriteObjectEnd();
+                            }
+                            writer.WriteArrayEnd();
```

Comment:
i'm curious why it's 1, followed by an array?  why not just do a single array?

Reply:
The vast majority of assemblies are single module. Multi-module assemblies are incredibly rare and not supported at all on .NET Core. I didn't want to complicate the output for that case.

Reply:
well it's a matter of perspective.  now you'll always have the `"secondaryModules": []` piece.  my recomenddation would be that you either `if` things so that isn't added if not needed.  Or just have it be always an array and hten not have the almost-always-pointless `secondaryModules` bit.  NOte: this is a big nit.  As you stated, this is not a format someone can depend on.  So it was just about potentially making the code to generate simpler (as well as reducing junk in the output).  If you prefer this, that's 100% fine.